### DISCUSSION TITLE
Any chance to allow `tar_read()` during a pipeline for custom formats? [help]

### DISCUSSION THREAD

**Discussion by edalfon** [ropensci/targets#1163] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

We just got this error message, after updating {targets} -we haven't updated for a while-:

> "target foo attempted to run targets::tar_read() during a pipeline, which is unsupported except when format %in% c("file", "file_fast") and repository == "local". This is because functions like tar_read() attempt to access or modify the local data store, which may not exist or be properly synced in certain situations. Also, please be aware that some functions like tar_make() and tar_destroy() should never run inside a target. Please find a different workaround for your use case.

And yes, we do call `tar_read()` within an Rmd file that builds a report, which is the exception as stated in the docs.

> So please do not call these functions from inside a target as part of a running pipeline. The only exception is literate programming target factories in the tarchetypes package such as tar_render() and tar_quarto().

But, we are doing that using targets tailored to our specific use case, which also use a custom target storage format, overriding the default read/write/marshal etc. So we have a target factory that generates some targets in the pipeline that ultimately render the Rmd file, but it does a bunch of stuff before that. That's why we are not simply using tar_render() or tar_quarto() and also, due to our custom format, we cannot set the format to 'file'.

So I was wondering if it would be possible to extend the exception when format is `format_custom` and not only c("file", "file_fast").

I guess I can see why these kind of guardrails are needed/important, imposing restrictions that prevent the users from doing stuff that could bite them later down the line. But I think it is also fair to assume that if users are going to the lengths of defining a custom format, they probably know what they are doing (or at least they should know, and if they don't, it's up to them if issues arise later on).

What do you think?

**Comment by edalfon** [comment]
I swear I searched before posting, but somehow missed this question https://github.com/ropensci/targets/discussions/1158 from [@MilesMac](https://github.com/MilesMcBain). It seems to me it is pretty much the same issue we are facing. And probably just allowing `tar_read` when you have a custom format would also do the trick for them.

**Reply by wlandau** [reply]
I'm not sure I see why it is necessary to disable the `tar_read()` guardrails in your case. Maybe there is a workaround. It depends on what exactly you are trying to do with the custom format.

#1158 is a different scenario. If `tar_read()` acts on a different data store than the pipeline currently running, then as @MilesMcBain said, it is an unambiguously safe case.

**Reply by edalfon** [reply]
thx as always for a super fast reply. 

You are right, there is a workaround for us and indeed https://github.com/ropensci/targets/discussions/1158 is not exactly the same case. But both cases have in common the use of a custom format. 

Although in our case `tar_read()` does not act on a different data store, we felt it was still save to do it. After all, what we are doing is just the same thing `tar_render()` would do. But we want to do some pre-processing before rendering the Rmd.

So I felt allowing some more leeway when you are writing custom targets and defining a custom format could be warranted. But again, I understand the importance of guardrails for `{targets}` and we can work around this restriction in our case. 

Thanks again and cheers!
